### ISSUE TITLE
Instance org gets wrong after session timeout

### ISSUE THREAD

**Issue by RonnyB71** [Altinn/app-lib-dotnet#94]
### Description of the bug

Wrong organization used for instance after session timeout

### Steps To Reproduce

1. Log on as a user with access to multiple organizations
2. Select one of the organizations to represent
3. Start filling out an instance
4. Leave the browser session open and let it time out
5. Re-logon and select a different organization
6. Observe that the instance is now connected to the last organization and not the organization used when starting

**Comment by tjololo** [comment]
Attempting a summary of the findings, as I understand them:
Instances or data connected to the instance are not transferred to other orgs so there is no data leaked to other orgs.
This is mainly a UX bug as the frontend displays the information present in the cookie and not the instance data.

| Action | Reportee in cookie | Reportee in instance data/metdata | Reportee shown in app frontend |
| ------ | ------------------- | ----------------------------- | ---------------------------- |
| Login and choose to represent OrgA | OrgA | N/A | N/A |
| Instantiate a new instance | OrgA | OrgA | OrgA |
| Session timeout, re-login represent OrgB | OrgB | OrgA | OrgB |
| Open new tab open inbox and choose to represent OrgC | OrgC | OrgA | OrgC |

Frontend should display the reportee set in instance metadata if it differs from what is set in the cookie. 

We need to decide if a user choose to represent OrgB and opens an app for OrgA, should we change and update the values in the users cookie or only change the value presented to the user while he/she is in the context of that instance?

**Comment by tjololo** [comment]
Another error scenario I didn't catch in my previous comment.
If the user follows a direct link to the instance when not logged in he will be redirected to id-porten for authentication with a redirect url back to the app, but apps uses hashroutes and the hash part of the route is not part of the redirect route. Therefor a new instance is created for the organisation/person the user chooses as reportee, even if it's the same as the one he previously had started an instance for.
Described in issue https://github.com/Altinn/app-frontend-react/issues/801
The issue described in this comment should be resolved when https://github.com/Altinn/app-frontend-react/issues/344 is implemented.

**Comment by RonnyB71** [comment]
It's like @tjololo sums up. Although not very user friendly, it's by-design and not a security issue. But we should fix sooner rather than later. I've grouped this, and related issues under https://github.com/orgs/Altinn/projects/39/views/13.

**Comment by pckofstad** [comment]
We are concerned that this might effect the [PBU project at Udir](https://github.com/Altinn/app-frontend-react/issues/397) as well. It would be good to have this fixed before April.

**Comment by magnusjerre** [comment]
This bug can affect the [PBU project at Udir](https://github.com/Altinn/app-frontend-react/issues/181) since there are schools with shared computers which means that multiple users will login on the same computer. We aren't guaranteed that they remember to actively log out after filling out the survey (or if a fire alarm were to ring and they switch computers on resuming the survey), meaning that the AltinnPartyId-cookie with the previous userid remains. When a new user logs in, some of the calls to the app fail and they are presented with a blank page. For them to get around the problem they would need to clear the cookie for AltinnPartyId then refresh the page, which we can't really expect of them to know.

I've added some screenshots reproducing the bug, note that the page on the left hand side is blank for the second login.

![cookiebug](https://user-images.githubusercontent.com/3714805/218683948-d94e9696-4826-49ed-8f1f-88108f030771.png)

**Comment by ivarne** [comment]
Isn't this really an issue in app-frontend-react that reports the results from the `/user` endpoint in the top right corner (which depends on the cookie), instead of showing the instance owner of the instance that is currently active?

As I understand the `AltinnPartyId` cookie, it is just a "suggestion" for what party to show the inbox for and to create new instances owned by. In all other cases you have `instance.Owner.PartyId` to specify the current party.

**Comment by pckofstad** [comment]
The problem for PBU at Udir is not that the wrong org is selected. The problem is that it seems to crash and give the user a white page of death.

**Comment by TheTechArch** [comment]
@ivarne  I believe this should be fixed with FE updates. FE should present the instance partyID, and we should consider to update party cookie if mismatch (like in Altinn 2)

**Comment by ivarne** [comment]
My little summary to see if I understand.

### Bug 1:
Front End (FE) currently uses the endpoint `/api/authorization/parties/current?returnPartyObject=true` and displays the result as `instanceOwner` in the upper right corner. This is wrong as this endpoint reads the `AltinnPartyId` cooke that might be out of sync, and FE should be updated to fetch the correct instance owner for the current instance instead of depending on the value in the cookie.

`parties/current` (and other `AltinnPartyId` dependent apis) should not be used for working in contexts where there exists an instance.

### Bug 2:
As I understand this, "UDIR" complains that frontend crashes if the current url `#/instance/{instanceOwner}/{instnaceGuid}` is pointing to an instanceOwner that the current user does not have any roles with (because the an instance was open, timed out, and another user logged in). This results in a correct `403 Forbidden` from backend and can be improved by showing an unauthorized screen, or just redirecting to the root of the app. In most usages, I think an unauthorized screen is probably the best option.

### Bug 3?:
Backend does not update the `AltinnPartyId` cookie in some specific cases where we might think that it should.
* The current authentication cooke does not have any roles with the party in the cookie. The code seems to try to do this, but I'm not confident.
* A user is fetching data about an instance not owned by the current `AltinnPartyId` cookie value. If frontend didn't fetch `parties/current` in parallel with the instance, this would be a possible (but bad) way to solve Bug1 in backend.

**Comment by FinnurO** [comment]
Prioritert med Udir. Foreslår et møte for å få oversikt.

**Comment by tjololo** [comment]
I think @ivarne summary is correct.
Bug 1: The endpoint/cookie should not be used to display who is the instanceOwner in the upper right corner. There are two scenarios where I think the endpoint/cookie should be used: When instantiation a new instance, and opening a stateless application. In all other cases the instanceOwner should be fetched from the instance information. BE cannot change the cookie as the user might have instances from different parties open in multiple tabs.

Bug 2: 403 seems like the correct responsecode. User is forbidden to open another users instanse. We need to investigate the 500 response from `/api/authorization/parties/current` in UDIRs responses

Bug 2: I think updating/changing cookies in apps is a anti-pattern. This would lead to an unknown number of altinn 3 apps with different code and versions of altinn app nugets updating the same cookies. I suspect this will result in a debugging nightmare down the road.
I think the correct place to update the cookie is in the authentication process. Once a user logs in the value in the cookie should be checked and removed if the newly logged in user cannot represent the partyId in the cookie.

**Comment by tjololo** [comment]
This will be resolved when 
* https://github.com/Altinn/altinn-authorization/issues/499 
* https://github.com/Altinn/app-lib-dotnet/issues/327
* https://github.com/Altinn/app-frontend-react/issues/1533
are closed
